Create 3 version of reports for 5 active Users per location..

shSachdeva opened the pull request.

shSachdeva committed:
Add Job to generate Report on Persisted Twitter Data

This application is added to create report of every day active user per location as per their tweeting activities.

shSachdeva committed:
Add Functionality to parse Json from Tweet Status

Tweet status are in payload in Json Format. To get all the status attributes as seperate fields, this Json is parsed into a Dataframe and required columns are filtered.

shSachdeva committed:
Generate Report for Twitter Analytics

Twitter report is generated with 3 different versions from
* SQL
* DataFrame
* DataSet API

shSachdeva committed:
Add function to save reports on HDFS

shSachdeva committed:
Merge branch 'TwitterReportingTask' of https://github.com/ipetrushin-gd/data-engineering-practice into TwitterReportingTask

ipetrushin-gd commented:
@shSachdeva ,
I see that you did put some effort to commit the changes in atomic fashion, but they are still not 100% atomic.

Please read [the article](https://medium.com/@fagnerbrack/one-commit-one-change-3d10b10cebbf), then analyze the PR commits and respond with an observation why the commit are still not atomic and how it could be achieved.

shSachdeva commented:
Order of commits is not correct :(
So, it can break the application. 
Need to commit utils first and then the application.
Also one change per commit rule is to be followed. But for new addition of files, does it mean part by part commit ?

shSachdeva committed:
Add file to save report to HDFS

This file is added to have a seperate api for saving report in HDFS instead of saving it through TweetsDataLoader.

shSachdeva committed:
Update saveReportToHdfs function API from TweetsDataLoader to TwitterReportWriter

shSachdeva committed:
Remove saveOutputToHdfs function from file

shSachdeva committed:
Add Input TestData to test Report Creation

shSachdeva committed:
Add expected TestData results to test generated report

shSachdeva committed:
Add Testcases to test that Dataframe with seperate fields is being created with a Json containing Parquet file

shSachdeva committed:
Add Testcases to test that reports are being created correctly in all three versions

shSachdeva committed:
Add testcase to confirm that report is being saved at specified path.

shSachdeva committed:
Drop the Column Frequency from generated reports

shSachdeva committed:
Revert "Add Testcases to test that reports are being created correctly in all three versions"

This reverts commit b745afd

shSachdeva committed:
Add Testcases to test that reports are being created correctly in all three versions

shSachdeva commented:
@ipetrushin-gd 
Realized an issue with implementation. Please refrain from review. Will update as soon as I fix it.

shSachdeva committed:
Add Logic to get Top 5 users per group each case

shSachdeva committed:
Correct name of the testcase relatively

shSachdeva commented:
@ipetrushin-gd 
I have fixed the issue. Just the DataSet version is not implemented completly in a typed-safe manner. I have done an attempt which is commented and it is giving an Encoder error for your review.

ipetrushin-gd commented:
>  I have done an attempt which is commented ...

how much time did you spend on the attempt?

shSachdeva commented:
Only this portion of DataSet in which we need to get top n rows per group from dataset in a typed safe manner is having an issue. Spent few hours on it. If you review and see that I am moving in correct direction, I can spend more time on it.

ipetrushin-gd commented:
the client expects our engineers to be able to operate in a self-sufficient manner. Please try to overcome the issue on your own.

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/JsonTranslator.scala:
```diff
@@ -0,0 +1,23 @@
+    import spark.implicits._
+
+    val unParsedDf = spark.read.parquet(path)
+    log.debug("=======> Stage 1: Twitter Data Read Start..")
```
log messages that specify order of an operation in a workflow  ("Stage #....") should be in a "flow controller" logic, not inside of a component.

shSachdeva replied:
Fixed..

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/TwitterReportGenerator.scala:
```diff
@@ -0,0 +1,60 @@
+    val query = "select location,id,count(id) as frequency from status group by location,id having frequency >= 10"
+    spark.sql(query).withColumn("Date",lit(current_date)).createOrReplaceTempView("filteredTable")
+
+    spark.sql("select location, id, Date from (select *, row_number() " +
```
please , use `"""..."""` for multi-line string formatting.

shSachdeva replied:
Fixed..

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/TwitterReportGenerator.scala:
```diff
@@ -0,0 +1,60 @@
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.expressions.scalalang.typed.{count => typedCount}
+
+object TwitterReportGenerator {
```
please apply "Strategy" pattern.

shSachdeva replied:
Was going through an article which explains how to compensate for Strategy Pattern in Scala. Have used that instead and pushed the code. Please see if that's okay or you want to implement it purely OOP way ?

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/TwitterReportingJob.scala:
```diff
@@ -0,0 +1,55 @@
+import org.apache.log4j.Logger
+import org.apache.spark.sql.SparkSession
+
+object TwitterReportingJob {
```
there are going to be many different reports.
please apply appropriate naming facilitating extension.

shSachdeva replied:
Fixed..

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/TwitterReportingJob.scala:
```diff
@@ -0,0 +1,55 @@
+
+    val fs = FileSystem.get(spark.sparkContext.hadoopConfiguration)
+    if(fs.exists(new Path(dataPath))) {
+      val reportSavePath = "twitter-analytics/reporting"
```
should be externalized to the config.

ipetrushin-gd replied:
there are going to be many different reports...

shSachdeva replied:
Fixed..

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/TwitterReportingJob.scala:
```diff
@@ -0,0 +1,55 @@
+      log.debug("=======> Stage 6: Saving Report on HDFS!")
+      TwitterReportWriter.saveReportToHdfs(reportFromSql, reportSavePath + "/sqlReport")
+      TwitterReportWriter.saveReportToHdfs(reportFromDf, reportSavePath + "/dataFrameReport")
+      TwitterReportWriter.saveReportToHdfs(reportFromDs, reportSavePath + "/dataSetReport")
```
no need to run every implementation in the single driver.
please design the code in a way that allows to run the driver with a particular implementation.

shSachdeva replied:
Are you suggesting every implementation goes through its own driver ?

ipetrushin-gd replied:
it's up to you how you decide to implement it.
the main point is - current approach has a bit of overhead.

shSachdeva replied:
Fixed..

ipetrushin-gd commented on src/test/resources/expectedResultFromReport.csv:
```diff
@@ -0,0 +1,6 @@
+"NewYork",101
+"NewYork",104
```
why only 2 users of 4 are expected for "NewYork" location?

shSachdeva replied:
Because only these IDs have frequency more than 10. So code picks any 5 or less entries which have been repeated more than 10 times.

ipetrushin-gd replied:
>  only these IDs have frequency more than 10.

that is fine, and I see it's being tested.


>  picks any 5 or less 

should also check "not more than 5" and "5"

ipetrushin-gd replied:
plus, it should pick  not "any", but **TOP**.

shSachdeva replied:
Fixed

ipetrushin-gd reviewed (changes requested):
@shSachdeva 
> I have fixed the issue. Just the DataSet version is not implemented completly

the tests cases in `TwitterReportGeneratorTest.scala`  don't seem to be validating what is expected per the original requirement.

shSachdeva committed:
Add code for report generation through Dataset typesafe manner

Report creation via Dataset implementation that was causing issue due to encoder has been fixed.

shSachdeva committed:
Wrap multi line SQL query in """ instead of +

shSachdeva committed:
Add configuration for path to save Report

shSachdeva committed:
Get configuration for Reports Save Path

shSachdeva commented:
Do you mean the Description of Testcases is not complying with requirements or there is something wrong with implementation ?

shSachdeva committed:
Update Testcase Description to imply complete requirement

A part of requirement that states that "only maximum of 5 or less users should be selected per location" was missing from testcase description. Testcases are modified accordingly.

shSachdeva committed:
Move Methods for report process into a new file

Report processing methods via all three methods have been moved here from TwitterReportGenerator to pass these algorithms into Strategy plan.

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/JsonTranslator.scala:
```diff
@@ -0,0 +1,23 @@
+import org.apache.log4j.Logger
+import org.apache.spark.sql.{DataFrame, SparkSession}
+
+object JsonTranslator {
```
the generic name implies that the object provides functionality for "working" with Json in general.
But in fact, a particular dataset gets processed.

E.g.:
`JsonTranslator#getFilteredDf` signature says about format and technology/framework, but not about actual business domain applicability. This is an anti-pattern in designing APIs. Please make it more specific, self-descriptive , reflecting **business** logic.

ipetrushin-gd commented on src/test/scala/com/gd/twitteranalytics/TwitterReportGeneratorTest.scala:
```diff
@@ -0,0 +1,47 @@
+
+  "TwitterReportGenerator " - {
+    "getReportWithDataFrameProcessing" - {
+      "should generate report with users that are occuring more than 10 times per location" in {
```
what about "top 5" part of the requirements?

shSachdeva committed:
Add Strategy to pass Report processing Alogs

Execute function is added where report processing algorithms can be passed instead of having different algos to be called

shSachdeva committed:
* Call execute functions and pass report processing algorithms to execute function instead of calling algos separately

*Configure save path for Reports

shSachdeva committed:
Update method getReportWithDataSetProcessing with corrections in columns.

shSachdeva committed:
Correct Typo in function name

shSachdeva committed:
Update Class name from TwitterReportGeneratorTest to ReportProcessorTest

shSachdeva committed:
Add Strategy to pass Report processing Alogs

Execute function is added where report processing algorithms can be passed instead of having different algos to be called

Move report processing Algos to ReportProcessor

shSachdeva committed:
Add Strategy to pass Report processing Alogs

Execute function is added where report processing algorithms can be passed instead of having different algos to be called

Move report processing Algos to ReportProcessor

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/TwitterReportGenerator.scala:
```diff
@@ -13,8 +13,8 @@ object TwitterReportGenerator {
 
-    spark.sql("select location, id, Date from (select *, row_number() " +
-      "OVER (PARTITION BY location ORDER BY frequency DESC) as rn  FROM filteredTable) tmp where rn <= 5")
+    spark.sql("""select location, id, Date from (select *, row_number()
```
please apply consistent and proper SQL formatting for all queries (https://www.sqlstyle.guide/#query-syntax)

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/TwitterReportGenerator.scala:
```diff
@@ -13,8 +13,8 @@ object TwitterReportGenerator {
-    spark.sql("select location, id, Date from (select *, row_number() " +
-      "OVER (PARTITION BY location ORDER BY frequency DESC) as rn  FROM filteredTable) tmp where rn <= 5")
+    spark.sql("""select location, id, Date from (select *, row_number()
+      OVER (PARTITION BY location ORDER BY frequency DESC) as rn  FROM filteredTable) tmp where rn <= 5""")
```
> query

what does it do?

> filteredTable

what does it contain?

> tmp

what does it contain?

such naming makes code harder to understand. Please make it more self-documenting.

shSachdeva replied:
Fixed..

ipetrushin-gd commented on src/test/scala/com/gd/twitteranalytics/TwitterReportGeneratorTest.scala:
```diff
@@ -11,7 +11,7 @@ class TwitterReportGeneratorTest extends FreeSpec with BeforeAndAfter with DataF
   "TwitterReportGenerator " - {
     "getReportWithDataFrameProcessing" - {
-      "should generate report with users that are occuring more than 10 times per location" in {
+      "should generate report with maximum of 5 users that are occuring more than 10 times per location" in {
```
the original requirement is finding top 5, not maximum 5. that makes difference.

shSachdeva replied:
Fixed..

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/ReportProcessor.scala:
```diff
@@ -0,0 +1,50 @@
+      where($"rn" <= 5).drop("rn").drop("frequency")
+  }
+
+  def getReportWithDataSetProcessing1(spark: SparkSession, tweetStatusDf: DataFrame):DataFrame = {
```
> getReportWithDataSetProcessing1

why is it "1" ?

shSachdeva replied:
slipped in unintentionally.. Fixed.

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/ReportProcessor.scala:
```diff
@@ -0,0 +1,50 @@
+import org.apache.spark.sql.functions.{count, current_date, lit, row_number}
+import org.apache.spark.sql.{DataFrame, Dataset, Encoders, SparkSession}
+
+object ReportProcessor {
```
two points:
1) I see that for some reason you call objects and classes "files":
> ... into a new file

To my mind, a "file" is not quite useful abstraction when we talk about OOP/FP design and implementaiton.

2) 
> ... methods  have been moved here from TwitterReportGenerator

I don't see TwitterReportGenerator changed. this is not what "moved" implies.

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/TwitterReportGenerator.scala:
```diff
@@ -1,51 +1,9 @@
-case class TwitterStatus(location:String,id:Long)
-case class Report(location:String,id:Long,frequency:Long)
\ No newline at end of file
+  def execute(reportProcessor: (SparkSession, DataFrame) => DataFrame,
```
the commit with change of the API breaks the build, since `TwitterReportingJob` logic was not adjusted accordingly.

shSachdeva replied:
This happened since I realized wrong order of commits and reverted those. But it had already been pushed. So it showed on my local copy and not on remote. I kind of got confused here :(

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/TwitterReportingJob.scala:
```diff
@@ -45,11 +47,7 @@ object TwitterReportingJob {
-      SparkSession.builder()
-                  .appName("TwitterKPIReport")
-                  .getOrCreate()
+      SparkSession.builder().appName("TwitterKPIReport").getOrCreate()
```
Again... There are going to be many different reports, many different KPIs. This fact should be taken into account right away, i.e. the design and naming should accommodate future extension (e.g. adding a new  report should not affect existing ones, lookup/search of a particular report should be straightforward)

shSachdeva replied:
Fixed.

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/TwitterReportingJob.scala:
```diff
@@ -26,13 +27,14 @@ object TwitterReportingJob {
       log.debug("=======> Stage 1: Twitter Data Read Complete !")
       val filteredStatusDf = JsonTranslator.getFilteredDf(spark, tweetsDataFrame)
```
`filteredStatusDf` doesn't give a clue what is inside of the Dataframe. What exactly has been filtered in/out?
Please make  the code more self-descriptive.

shSachdeva replied:
Fixed.

ipetrushin-gd reviewed (changes requested):
P.S.
the comment subject format is invalid: https://github.com/ipetrushin-gd/data-engineering-practice/pull/4/commits/177c3635cea238929d30b2c32f891face2189fcb

shSachdeva committed:
Delete TwitterReportingJob to facilitate dedicated drivers for individual jobs

shSachdeva committed:
Rename JsonTranslator to ReportInputDataParser

* JsonTranslator object is deleted
* Object ReportInputDataParser is created from JsonTranslator and moved in to reports package
* Relevent Test object has also been updated

shSachdeva committed:
Move object TwitterReportGenerator to package reports

shSachdeva committed:
Move object TwitterReportWriter to package reports

* TwitterReportWriter is moved to package reports
* Relevent Test object is modified accordingly

shSachdeva committed:
Move object ReportProcessor to package reports

* ReportProcessor is moved to package reports
* Relevant Test object is modified accordingly

shSachdeva committed:
Update Input Test Data

Dataset is populated with more entries to test all three scenarios that is 5, more than 5 and less than 5 frequency.

shSachdeva committed:
Update expected test resultData

* expected test result data is modified to show that only 5 or less top entries make to report and more than 5 entries are discarded.

shSachdeva committed:
Add Dedicated Drivers for each version of ActiveUsers report

shSachdeva committed:
Revert "Add Dedicated Drivers for each version of ActiveUsers report"

This reverts commit 522625d

shSachdeva committed:
Add Trait TwitterReport for ActiveUser reports of 3 versions

shSachdeva committed:
Add Dedicated Drivers for each version of ActiveUsers report

shSachdeva committed:
Update Variable names to imply Logic

shSachdeva committed:
Reformat SQL queries

SQL queries are formatted in ReportProcessor object and related testcases are updated accordingly with correct column name.

shSachdeva committed:
Update filename to be in sync with object name

shSachdeva committed:
Add configuration file for Report Date Retention Script

shSachdeva committed:
Add Bash Script to purge data older than 7 days

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/ReportInputDataParser.scala:
```diff
@@ -0,0 +1,19 @@
+
+  def getPayloadStatusAsDataset(spark:SparkSession,path:String):DataFrame ={
+    import spark.implicits._
+    val unParsedDf = spark.read.parquet(path)
```
NOTE: I suppose it should be formatted as`unparsedDf`

ipetrushin-gd replied:
There is a dedicated Tweet model, right? Please read with the model, so it's clear what schema being processed.

shSachdeva replied:
Tweet Model has attributes event_date and payLoad. Here we are directly reading from path that includes event_date partition for report creation, so can't use Tweet model.

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/ReportInputDataParser.scala:
```diff
@@ -0,0 +1,19 @@
+
+object  ReportInputDataParser{
+
+  def getPayloadStatusAsDataset(spark:SparkSession,path:String):DataFrame ={
```
could you clarify please, why `AsDataset` if `DataFrame` is returned?

shSachdeva replied:
Fixed..

ipetrushin-gd commented on src/test/scala/com/gd/twitteranalytics/ReportInputDataParser.scala:
```diff
@@ -1,13 +1,13 @@
 import org.scalatest.{BeforeAndAfter, FreeSpec}
 
-class JsonTranslatorTest extends FreeSpec with DataFrameSuiteBase with BeforeAndAfter{
+class ReportInputDataParserTest extends FreeSpec with DataFrameSuiteBase with BeforeAndAfter{
```
according to convention, it should be `Spec` not `Test`.

shSachdeva replied:
Can you please send a link for this naming convention..

ipetrushin-gd replied:
check out the naming in the official docs: http://www.scalatest.org/user_guide/selecting_a_style

ipetrushin-gd reviewed (changes requested):
The different drivers look fine, but they're a lot of copy past which is an anti-pattern.
please generalize/reuse common parts.

P.S.
Corresponding test cases for strategies also have a lot of copy past...

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/TwitterReport.scala:
```diff
@@ -0,0 +1,25 @@
+  }
+
+  def setSparkSession(name:String):SparkSession = {
+    SparkSession.builder.appName(name).master("local[*]")getOrCreate
```
why  is it `local[*]`?

ipetrushin-gd replied:
isn't it an environment dependent parameter?

shSachdeva replied:
It slipped in un-intentionally..

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/ReportProcessor.scala:
```diff
@@ -0,0 +1,65 @@
+import org.apache.spark.sql.functions.{count, current_date, lit, row_number}
+import org.apache.spark.sql.{DataFrame, Dataset, Encoders, SparkSession}
+
+object ReportProcessor {
```
there are going to be many different reports!

shSachdeva replied:
Fixed..

shSachdeva committed:
Fix getReportWithDataSetProcessing Testcase Failure

Testcase related to DataSet version of top5 active user was failing because of sort not happening.

shSachdeva committed:
Update ReportProcessor object to ActiveUserReportProcessor

Name of object ReportProcessor is updated to ActiveUserReportProcessor and related testcase names and references are updated accordingly.

shSachdeva committed:
Remove MasterUrl from sparkSession

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/TwitterReportWriter.scala:
```diff
@@ -0,0 +1,11 @@
+object TwitterReportWriter {
+
+  def saveReportToHdfs(inputDataSet: DataFrame, savePath:String)={
+    inputDataSet.write.partitionBy("Date").
```
> Date

should follow SQL naming convention for columns.

shSachdeva replied:
Fixed..

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/TwitterReportGenerator.scala:
```diff
@@ -0,0 +1,11 @@
+
+object TwitterReportGenerator {
+
+  def execute(f: (SparkSession, DataFrame) => DataFrame, spark: SparkSession, tweetStatusDf: DataFrame) = f(spark, tweetStatusDf)
```
why there are two `execute` functions?
it even causes compilation error. 
Please run tests before committing and pushing to a remote branch.

shSachdeva replied:
Is it showing two execute functions in your copy ? At branch, I can see only one..

ipetrushin-gd replied:
well.. this is from the branch: https://github.com/ipetrushin-gd/data-engineering-practice/blob/dbbb4bb45e9a6f39a51e64f860d555aeba5e6aed/src/main/scala/com/gd/twitteranalytics/TwitterReportGenerator.scala

shSachdeva replied:
Wow !!
Sorry, this is some git glitch.. because it would clearly show me a compilation
 error if I would try to commit file with two same methods.
Fixing it rite away..

ipetrushin-gd commented on src/main/resources/configForReports.conf:
```diff
@@ -0,0 +1,5 @@
+report_path=twitter-analytics/reporting/ActiveUsers
```
> configForReports.conf
> report_path
> dfReportPath
...


there are going to be many different reports!

shSachdeva replied:
Fixed..

ipetrushin-gd commented:
@shSachdeva ,
please reply to a reviewer's comments once they have been addressed.
your answer can be as simple as "fixed.". It simplifies review.

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/TwitterReport.scala:
```diff
@@ -0,0 +1,25 @@
+
+  val today = java.time.LocalDate.now
+  val dataPath = AppConfigReader.getAppConfigurables(0) + "/event_date=" + today
+  val reportSavePath = AppConfigReader.getAppConfigurables(3) + "/ActiveUsers"
```
**once once once again**: there are going to be many different reports!

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/ActiveUserDatasetReport.scala:
```diff
@@ -0,0 +1,33 @@
+    val fs = FileSystem.get(spark.sparkContext.hadoopConfiguration)
+
+    if (fs.exists(new Path(dataPath))) {
+      log.debug("=======> Building Twitter ActiveUser Report for date " + today)
```
for backfilling or re-processing purposes, it should be possible to run a report for any particular date in the past.

shSachdeva replied:
Okay.. so current logic is written on the requirement that report will be run on a daily basis on a daily data... To implement reporting for re-processing purpose, how we are supposed to get past date on which report is to be run ? Through Config ?

ipetrushin-gd replied:
> report will be run on a daily basis on a daily data...

it implies a scheduling requirement.
In general, backfilling and re-processing activities are parts of data engineering process, that I assumed you was familiar with.

ipetrushin-gd replied:
> how we are supposed to get past date on which report is to be run ? Through Config ?

usually it's done via arguments or config, yes.

shSachdeva replied:
Fixed.

shSachdeva committed:
Fix Git Remote branch with two same method definitions..

shSachdeva committed:
Delete duplicate Object

Object TwitterReportGenerator is part of reports package instead.

shSachdeva committed:
Update config properties name to reflect corresponding report

* Property names are updated to reflect that these properties relate to activeUsers report.
* Related script file purgeOldReport.sh has been updated accordingly.

shSachdeva committed:
Update script name from purgeOldReportData to purgeOldAciveUsersReportData

shSachdeva committed:
Update method name getPayloadStatusAsDataSet to getPayloadStatusAsDataFrame in object ReportInputDataParser

* Occurance of refactoring method name is updated in corresponding objects and testcases..

shSachdeva committed:
Update column name "Date" to "date" as per SQL naming conventions

shSachdeva committed:
Update report related configuration from application to reporting

* report related configuration is removed from application to reporting in application.conf
* Corresponding configuration read is updated in AppConfigReader
* Related changes to configuration is updated in twitter report trait and objects.

shSachdeva committed:
Add Trait to remove duplicate code from test strategies

* ReportProcessorTestSuite is added to work as an interface to provide common code to all report processor testcases
* ActiveUserReportProcessorTest testcases are updated accordingly.

shSachdeva committed:
Remove common code from ActiveUser report drivers

* Common code has been removed from report driver objects and moved into trait TwitterReport for re-usability purpose.

shSachdeva committed:
Revert "Add Trait to remove duplicate code from test strategies"

This reverts commit b956fad

shSachdeva committed:
Update date column with eventDate of data on which report is being built

* Instead of using today's date to use as a field in 'date' column, use the date of data on which report is being built.
* Relevant objects and testcases are updated accordingly

shSachdeva committed:
Remove MasterUrl from sparkSession

shSachdeva commented:
> P.S.
> Corresponding test cases for strategies also have a lot of copy past...

In Testcase strategies, we are using DataFrameSuiteBase. This has spark as a lazy sparkSession which can only gets initialized inside a test case otherwise gives null value. So if I try to move common code like inputDf and expectedDf inside a trait and use in strategies
, I get an NPE. So I have to have Dataframes inside each testcase.

shSachdeva committed:
Merge remote-tracking branch 'origin/TwitterReportingTask' into TwitterReportingTask

ipetrushin-gd commented on Deployment/purgeOldReportData.sh:
```diff
@@ -13,10 +13,10 @@ echo "=====> Older Data is from date : $past_date"
 echo "=====> $current_date: INFO: Scanning reports Data for retention.."
 
-echo "=====> INFO: Purging reports older than $past_date from $report_path....."
+echo "=====> INFO: Purging reports older than $past_date from $activeUserReport_path....."
```
> purgeOldReportData.sh

the procedure is related to retention management.
please use this fact for naming more abstract "purge old data".

ipetrushin-gd replied:
provided that there are going to be many different reports  (which is a fact that I've stated multiple times), could  you plan for extension beforehand? i.e. please make retention management logic reusable for any dataset and for any period?

shSachdeva replied:
I have already updated the file name.. This is an outdated file.. New name is purgeOldActiveUsersReportData.sh
And I was thinking of making different scripts for different reports.. Do you want me to create one single script which will do data retention for all kind of reports ?

shSachdeva replied:
I have updated the script to be used as a generic script for any kind of data retention. Script name has been updated accordingly.

ipetrushin-gd commented on src/main/resources/configForReports.conf:
```diff
@@ -1,5 +1,5 @@
-report_path=twitter-analytics/reporting/ActiveUsers
+activeUserReport_path=twitter-analytics/reporting/activeUsers
```
> configForReports.conf

could you clarify, please: Is your idea to store define configs for different repots in the same file?

shSachdeva replied:
Yes, this file is supposed to contain configuration for all different report scripts.

ipetrushin-gd replied:
ok, thanks for clarifying.

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/ReportInputDataParser.scala:
```diff
@@ -4,16 +4,15 @@ import org.apache.spark.sql.{DataFrame, SparkSession}
 object  ReportInputDataParser{
 
-  def getPayloadStatusAsDataset(spark:SparkSession,path:String):DataFrame ={
+  def getPayloadStatusAsDataFrame(spark:SparkSession,path:String):DataFrame ={
```
As you can see, subject of the commit is too long, so it doesn't get displayed properly.
please follow the guide for proper commits. (the guide from the reading list)

shSachdeva replied:
This particular commit message is inserted by GitHub

ipetrushin-gd commented on src/main/resources/application.conf:
```diff
@@ -2,5 +2,8 @@ application {
+}
+reporting {
   reportSavePath = "twitter-analytics/reporting"
+  eventDateOfDataForReport = "2018-11-02"
```
so, do we need to package/build different artifact for each date?

shSachdeva replied:
Fixed. This particular property which is more dynamic in nature is moved to an external config file that is external to jar.

ipetrushin-gd commented on src/test/scala/com/gd/twitteranalytics/ReportProcessorTestSuite.scala:
```diff
@@ -0,0 +1,16 @@
+import org.apache.spark.sql.functions.{current_date, lit}
+import org.scalatest.FreeSpec
+
+trait ReportProcessorTestSuite extends FreeSpec with DataFrameSuiteBase{
```
is this trait supposed to be used only for active users report?

shSachdeva replied:
Sorry, but this is an outdated file. It's been removed.

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/TwitterReport.scala:
```diff
@@ -9,8 +11,22 @@ trait TwitterReport {
+  }
+
+  def getInputDataForActiveUserReport(dataPath:String,spark:SparkSession,log:Logger):DataFrame = {
+    log.debug("=======> Stage 2: Reading Data for Twitter Active User Report")
```
Again... the order based logging should be in object that actually calls the functions, because it's the object who sets up the order.

shSachdeva replied:
Fixed.

shSachdeva committed:
Update purgeOldActiveUsersReportData.sh to reportDataRetention.sh

* purgeOldActiveUsersReportData is updated to be more generic and take any kind of report dataset for retention.
* Related Configuration file has been updated accordingly.

ipetrushin-gd commented on src/main/resources/configForReports.conf:
```diff
@@ -2,4 +2,8 @@ activeUserReport_path=twitter-analytics/reporting/activeUsers
\ No newline at end of file
+activeUserDsReportPath=${activeUserReport_path}/dataSetReport
+
+retentionPeriod=7
```
the config is dedicated to spark job building report.
retention management doesn't get handled by the job, another/separate application is responsible for the procedure. another application - separate config.

ipetrushin-gd replied:
> reportDataRetention.sh

please you imperative naming.

shSachdeva replied:
Name of script updated to reportRetentionManager.sh
And this configuration is related not to spark building report. It is related to retention management script. Name of conf updated to configForReportsRetention

ipetrushin-gd replied:
> Name of script updated to reportRetentionManager.sh

are you sure that it is imperative now?

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/ActiveUserReportProcessor.scala:
```diff
@@ -0,0 +1,65 @@
+                                      GROUP BY location,
+                                               id
+                                      HAVING frequency >= 10"""
+    spark.sql(activeUsersIdAndLocation).withColumn("date", lit(reportDate)).createOrReplaceTempView("activeUsers")
```
please apply the following formatting for readability of such long sequences:
```scala

spark
    .sql(activeUsersIdAndLocation)
    .withColumn("date", lit(reportDate))
    .createOrReplaceTempView("activeUsers")

```

shSachdeva replied:
Fixed..

shSachdeva committed:
Remove duplicate code from testcases

Common Datasets needed for all test strategies have been moved to before{} that setsup the data for each testcase.

shSachdeva committed:
Make eventDate for reports an External configuration

* New object ExternalConfigLoader added to read properties from a file kept in conf folder at jarPath.
* References for reportDate has been updated in updated in related objects.

shSachdeva committed:
Move order based Logging in driver objects

shSachdeva committed:
Reformat code for better readability

shSachdeva committed:
Update script name to reportRetentionManager

shSachdeva committed:
Update configuration file name

shSachdeva committed:
Remove MasterUrl from sparkSession

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/ActiveUserDataFrameReport.scala:
```diff
@@ -1,17 +1,23 @@
+      log.debug("=======> Stage 1: Reading Data for Twitter Active User Report...")
       val inputDataForReport = getInputDataForActiveUserReport(dataPath, spark,log)
+
+      log.debug("=======> Stage 1: Creating report for Twitter Active Users...")
```
should it be stage **2**?

shSachdeva replied:
Fixed.

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/ExternalConfigLoader.scala:
```diff
@@ -0,0 +1,21 @@
+
+import com.gd.twitteranalytics.util.AppConfigReader
+
+object ExternalConfigLoader {
```
not sure if this idea will work with a workflow orchestrator. 
typically an orchestrator accepts a packaged artifacts bundle for execution of a workflow, so the "external config" should be placed inside of the bundle. In this case, re-packaging of the bundle is required in order to update the date.

Explicit Command-line argument or system property are more viable options.

Have you tested this approach with Ozzie? Or may be you have done similar things in your past experience with other workflow orchestrators?

shSachdeva replied:
I haven't experimented external configuration setup with workflow managers yet.
I can try here in cluster setup. But the issue is that I am still stuck up with no access to HUE. I have had multiple requests for this.. but no success..

ipetrushin-gd replied:
ok. please add a TODO to check if the config handling  is feasible for deployment to the orchestrator.

shSachdeva replied:
Done.

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/ActiveUserDataFrameReport.scala:
```diff
@@ -0,0 +1,29 @@
+import ActiveUserReportProcessor._
+import TwitterReportGenerator._
+
+object ActiveUserDataFrameReport extends TwitterReport {
```
There are still copy-pasts in the TwitterReport implementations...

ipetrushin-gd replied:
What are principal differences between the implementations that you see?

shSachdeva replied:
> There are still copy-pasts in the TwitterReport implementations...

Can you please pin point this one ?

ipetrushin-gd replied:
Don't get me wrong, I'd like to help you here, but it takes one just to look and compare three different files and find similarities/repeated patterns - which is a pretty straightforward task.
So, please, put a bit more effort and "surprise" me with your design and refactoring skills :)

shSachdeva committed:
Correct the order of Logs Sequence

shSachdeva committed:
Add a TODO note for fesability check with orchestrator

shSachdeva committed:
Modify spark-submit script to choose one of the job to run

* Script is modified such as user can choose which script is required to run. TwitterStreamingJob or TwitterReportingJob.

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/ExternalConfigLoader.scala:
```diff
@@ -5,6 +5,7 @@ import java.util.Properties
 
 import com.gd.twitteranalytics.util.AppConfigReader
 
+//TODO Check the external configuration feasability with deployment to Oozie workflow manager..
```
NOTE: I noticed that you like to put double dots in the end of sentences - which is ok for personal/informal communication, but it doesn't seem to be ok for code documentation and comments, since they should be more formal.

shSachdeva replied:
Fixed.

shSachdeva committed:
Move duplicate code from ActiveUser report drivers to TwitterReport trait

shSachdeva committed:
Update TODO comment

ipetrushin-gd commented:
@shSachdeva 
subject of [the commit](
https://github.com/ipetrushin-gd/data-engineering-practice/pull/4/commits/6d11b33df1dc823b00100237124dbb3feee37627) is bit too long, so it doesn't get displayed properly.
please follow the guide for proper commits. (the guide from the reading list)

ipetrushin-gd commented on Deployment/spark-submit.sh:
```diff
@@ -2,6 +2,34 @@
+            echo "====> Running TwitterStreaming Application...."
+            break
+            ;;
+        "TwitterReportingJob")
```
does this job exist at all?

ipetrushin-gd commented on src/main/scala/com/gd/twitteranalytics/reports/TwitterReport.scala:
```diff
@@ -44,4 +44,10 @@ trait TwitterReport {
     saveReportToHdfs(report, activeUserReportSavePath + reportType)
   }
+
+  def stopJob(spark:SparkSession,log:Logger) = {
```
is it everything that you have found reusable?
